src/desktop/popup/manager.rs

```diff
@@ -56,24 +60,62 @@ impl PopupManager {
         }
     }
 
-    fn add_popup(&mut self, popup: PopupKind) -> Result<(), DeadResource> {
-        let mut parent = popup.parent().unwrap();
-        while get_role(&parent) == Some(XDG_POPUP_ROLE) {
-            parent = with_states(&parent, |states| {
-                states
-                    .data_map
-                    .get::<Mutex<XdgPopupSurfaceRoleAttributes>>()
-                    .unwrap()
-                    .lock()
-                    .unwrap()
-                    .parent
-                    .as_ref()
-                    .cloned()
-                    .unwrap()
-            })?;
+    /// TODO: Doc
+    pub fn grab_popup(&mut self, popup: PopupKind) -> Result<PopupGrab, PopupGrabError> {
+        let parent = popup.parent().unwrap();
+        let root = find_popup_root_surface(&popup)?;
+
+        let toplevel_popups = with_states(&root, {
+            let root = root.clone();
+            move |states| {
+                let toplevel_popups = ToplevelPopupGrab::new(root);
+                if states.data_map.insert_if_missing(|| toplevel_popups.clone()) {
+                    self.popup_grabs.push(toplevel_popups);
+                }
+                let toplevel_popups = states.data_map.get::<ToplevelPopupGrab>().unwrap().clone();
+                if !toplevel_popups.alive() {
+                    self.popup_grabs.push(toplevel_popups.clone());
+                }
+
+                toplevel_popups
+            }
+        })?;
+
+        // The parent of a grabbing popup must either be another xdg_popup with an active explicit grab, or an xdg_popup or xdg_toplevel, if there are no explicit grabs already taken.
+        let parent_role = get_role(&parent);
+
+        if parent_role != Some(XDG_TOPLEVEL_ROLE) && parent_role != Some(XDG_POPUP_ROLE) {
+            root.as_ref()
+                .post_error(
+                    wayland_protocols::xdg_shell::server::xdg_popup::Error::InvalidGrab as u32,
+                    "The parent of a popup grab must be either another xdg_popup with an active grab or and xdg_toplevel".to_string());
         }
 
-        with_states(&parent, |states| {
+        // Spec: If the parent is a grabbing popup which has already been dismissed, this popup will be immediately dismissed. If the parent is a popup that did not take an explicit grab, an error will be raised.
+        if parent_role == Some(XDG_POPUP_ROLE) {
+            if toplevel_popups.is_dismissed(&parent) {
+                popup.send_done();
+                return Err(PopupGrabError::ParentDismissed);
+            }
+
+            if !toplevel_popups.has_grab(&parent) {
+                root.as_ref()
+                .post_error(
+                    wayland_protocols::xdg_shell::server::xdg_popup::Error::InvalidGrab as u32,
+                    "The parent of a popup grab must be either another xdg_popup with an active grab or and xdg_toplevel".to_string());
+                return Err(PopupGrabError::InvalidGrab);
```

Comment:
The way I read the spec, it can be a popup without an explicit-grab, as long as there are no other grabs happening. (If there are, *then* the parent needs to be specifically the one-and-only popup with that grab. This does not seem to account for that?

Reply:
To be honest i am quiet confused by the spec.

At first it states

> The parent of a grabbing popup must either be an xdg_toplevel surface or another xdg_popup with an explicit grab. If the parent is another xdg_popup it means that the popups are nested, with this popup now being the topmost popup.

then shorty after

> The parent of a grabbing popup must either be another xdg_popup with an active explicit grab, or an xdg_popup or xdg_toplevel, if there are no explicit grabs already taken.

and near the end

> If the parent is a grabbing popup which has already been dismissed, this popup will be immediately dismissed. If the parent is a popup that did not take an explicit grab, an error will be raised.

Reply:
Very much understandable

> [...] or another xdg_popup with an explicit grab.  [...]


+1 for only xdg-popups with an explicit grab being valid targets

> [...]  either be another xdg_popup with an active explicit grab, or an xdg_popup or xdg_toplevel, if there are no explicit grabs already taken.


Clear point for either xdg_toplevel or xdg_popup, if no grabs exist, **or** an xdg_popup with a grab.

> [...] If the parent is a popup that did not take an explicit grab, an error will be raised.


So now we get an error, if the parent popup has no grab.. So +1 for only grabbed popups being valid parents.

It seems to contradict itself, so I would argue, that we should open an issue https://gitlab.freedesktop.org/wayland/wayland-protocols.

For the implementation we should probably follow what other compositors are doing right now?